### PR TITLE
Don't download tarballs if a local git repo found

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
Don't download tarballs if a local git repo found, to save time or avoid modify the makefile in local, make it easy to intergrate nuttx with customized third party software.
## Impact
Refactor only
## Testing
CI and local machine.

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: crypto/libtomcrypt/Makefile [hunk 1]
```diff
@@ -476,10 +476,13 @@ $(LIBTOMCRYPT_UNPACKNAME): $(LIBTOMCRYPT_ZIP)
 	$(Q) patch -p0 < ltcrypt-flush.patch # Missing stdout flush in demo app
 	$(Q) touch $(LIBTOMCRYPT_UNPACKNAME)
 
+# Download and unpack tarball if no git repo found
+ifeq ($(wildcard $(LIBTOMCRYPT_UNPACKNAME)/.git/HEAD),)
```
Comment (xiaoxiang781216) [inline crypto/libtomcrypt/Makefile:480]: it's enough to check .git folder?

Comment (no1wudi) [reply crypto/libtomcrypt/Makefile:480]: Yes, updated now

File: interpreters/wamr/Makefile [hunk 2]
```diff
@@ -48,12 +48,16 @@ $(WAMR_UNPACK): $(WAMR_TARBALL)
 	$(Q) echo "Unpacking $(WAMR_TARBALL) to $(WAMR_UNPACK)"
 	$(Q) unzip $(WAMR_TARBALL)
 	$(Q) mv wasm-micro-runtime-$(WAMR_VERSION) $(WAMR_UNPACK)
-	$(Q) touch $(WAMR_UNPACK)
```
Comment (xiaoxiang781216) [inline interpreters/wamr/Makefile:51]: why remove

File: crypto/libtomcrypt/Makefile [hunk 3]
```diff
@@ -476,10 +476,13 @@ $(LIBTOMCRYPT_UNPACKNAME): $(LIBTOMCRYPT_ZIP)
 	$(Q) patch -p0 < ltcrypt-flush.patch # Missing stdout flush in demo app
 	$(Q) touch $(LIBTOMCRYPT_UNPACKNAME)
 
+# Download and unpack tarball if no git repo found
+ifeq ($(wildcard $(LIBTOMCRYPT_UNPACKNAME)/.git),)
```
Comment (xiaoxiang781216) [inline crypto/libtomcrypt/Makefile:480]: should we guard line 466-477 too?

File: testing/ltp/Makefile [hunk 4]
```diff
@@ -132,9 +132,12 @@ $(LTP_UNPACK):
 	$(Q) git -C $(LTP_UNPACK) am < 0001-Fix-static-struct-warning.patch
 	$(Q) git -C $(LTP_UNPACK) am < 0002-Use-ifdef-instead-of-if-for-__linux__.patch
 
+# Download and unpack tarball if no git repo found
+ifeq ($(wildcard $(LTP_UNPACK)/.git),)
```
Comment (xiaoxiang781216) [inline testing/ltp/Makefile:136]: there are other 3rd party project not change(e.g. system/argtable3.

File: interpreters/wamr/Makefile (hunk 2 above)
Comment (no1wudi) [reply interpreters/wamr/Makefile:51]: Restored

File: testing/ltp/Makefile (hunk 4 above)
Comment (no1wudi) [reply testing/ltp/Makefile:136]: More other project updated.

File: crypto/libtomcrypt/Makefile (hunk 3 above)
Comment (no1wudi) [reply crypto/libtomcrypt/Makefile:480]: I think no need for that ? Small guard range is clear to understand and these dangling targets have no side effects for system.

File: mlearning/darknet/Makefile [hunk 5]
```diff
@@ -79,12 +79,14 @@ darknet.zip:
...
+# Download and unpack tarball if no git repo found
+ifeq ($(wildcard darknet/.git),)
 context:: darknet.zip
 
 distclean::
 	$(call DELDIR, darknet)
 	$(call DELFILE, darknet.zip)
-
+endif
 endif
```
Comment (xiaoxiang781216) [inline mlearning/darknet/Makefile:90]: let's remove ifeq ($(CONFIG_DARKNET_YOLO),y)/endif?

File: netutils/cjson/Makefile [hunk 6]
```diff
@@ -54,6 +56,8 @@ $(CJSON_UNPACKNAME): $(CJSON_TARBALL)
 	@echo "Unpacking: $(CJSON_TARBALL) -> $(CJSON_UNPACKNAME)"
 	$(Q) $(UNPACK) $(CJSON_TARBALL)
 	$(Q) touch $(CJSON_UNPACKNAME)
+else
```
Comment (xiaoxiang781216) [inline netutils/cjson/Makefile:59]: why add else?

File: system/embedlog/Makefile [hunk 7]
```diff
@@ -218,6 +220,9 @@ $(EMBEDLOG_SOURCES): $(EMBEDLOG_TARBALL)
 	$(Q) $(call DELDIR, $@)
 	$(Q) $(UNPACK) $<
 	$(Q) touch $@
+else
+$(EMBEDLOG_SOURCES):
```
Comment (xiaoxiang781216) [inline system/embedlog/Makefile:224]: don't need

File: testing/unity/Makefile [hunk 8]
```diff
@@ -53,6 +55,9 @@ $(UNITY_UNPACKNAME): $(UNITY_TARBALL)
 	@echo "Unpacking: $(UNITY_TARBALL) -> $(UNITY_UNPACKNAME)"
 	$(Q) $(UNPACK) $(UNITY_TARBALL)
 	$(Q) touch $(UNITY_UNPACKNAME)
+else
+$(UNITY_UNPACKNAME):
```
Comment (xiaoxiang781216) [inline testing/unity/Makefile:59]: don't need else clause

File: interpreters/quickjs/Makefile [hunk 9]
```diff
@@ -87,7 +92,9 @@ clean::
 	$(Q) test ! -d $(QUICKJS_UNPACK) || $(MAKE) -C $(QUICKJS_UNPACK) clean
 
 distclean::
+ifeq ($(wildcard $(QUICKJS_UNPACK)/.git),)
```
Comment (xiaoxiang781216) [inline interpreters/quickjs/Makefile:95]: move distclean into ifeq clause at line 80

**Comment by xiaoxiang781216** [comment]
@no1wudi should we do the similar change to nuttx repo too?

File: crypto/libtomcrypt/Makefile (hunk 3 above)
Comment (xiaoxiang781216) [reply crypto/libtomcrypt/Makefile:480]: Ok.

**Comment by mlyszczek** [comment]
Good idea but I think it's a bad approach. Right now, all packages hardcode their download code, they simply call "wget $(URL)", which makes it basically impossible to provide own tarballs (most common use case would be to protect yourself from upstream link going offline).

I think the right approach would be to create function that would take 2 arguments, base URL (like http://domain.com/distfiles) and filename (libpackage-1.0.0.tar.gz). Kconfig would hold cache url (or urls). Then, said download function would first try to get $filename from cache url(s) and then, once file could not be found, upstream URL from 1st argument would be used.

This will unify package download, and will limit clutter in app Makefiles.

**Comment by xiaoxiang781216** [comment]
@mlyszczek This change doesn't forbid we unify the download process to a common place. Let's improve the system step by step.

**Comment by mlyszczek** [comment]
Sure, but I still think proposed implementation is suboptimal, it adds unnecessary clutter and maintanence to all 3rd party apps.

New download function could also support git links with specified hash/tag. Idea is very good, but proposed implementation is simply a workaround, a band-aid, not a solution.

File: mlearning/libnnablart/Makefile [hunk 10]
```diff
@@ -115,11 +115,14 @@ nnabla.zip:
...
 	$(Q) mv nnabla-c-runtime-$(CONFIG_NNABLA_RT_VER) nnabla-c-runtime
 
+# Download and unpack tarball if no git repo found
+ifeq ($(wildcard nnabla-c-runtime/.git),)
 context:: nnabla.zip
 
 distclean::
 	$(call DELDIR, nnabla-c-runtime)
 	$(call DELFILE, nnabla.zip)
+endif
```
Comment (xiaoxiang781216) [inline mlearning/libnnablart/Makefile:125]: line 127 can be removed too.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**